### PR TITLE
tap: Buffered Admin Tap Filter Implementation

### PR DESCRIPTION
**Pull request by DavidPeet8**
**Commit Message:**
This change introduces a new output sink type for admin `/tap` requests which buffers traces internally before responding to the client.

This sink is best used to collect traces for requests that are frequently matched, or to work around system limitations such as improper support for streaming HTTP responses.

Contributed by: @Uber
Signed-off-by: David Peet <davidpeet@tutanota.com>

**Risk Level:** Medium
**Testing:** Additional integration tests created for the tap filter
**Docs Changes:** Additional description of buffering options added to tap filter documentation 
**Release Notes:** n/a
**Platform Specific Features:** n/a
**Fixes:** #19727

### CODE REVIEW

**Comment by jmarantz** [comment]
I had a quick scan and this looks great at first glance, but I am in the middle of perf and won't get to this right away. If this can wait a few days I'm happy to review, otherwise consider re-assigning.

**Comment by DavidPeet8** [comment]
No worries, this can wait a few days!

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:162]: should we do something here with the failed request, like emit an HTTP status, maybe 408 or 410, and maybe detail in the response buffer also?

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:181]: Some duplicate code with above; consider a helper class, maybe an intermediate parent in the class hierarchy?

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:220]: please use an explicit type in this case; as a reader of this code I can't tell from immediate context what type this will be. Envoy's style guide refs Google's style guide on this: https://google.github.io/styleguide/cppguide.html#Type_deduction

In particular I wondered if that should be a const ref.

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:222]: const

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:223]: const uint64_t timeout_ms

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:234]: I haven't worked that much with weak_ptr but should this be passed by const ref?

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:263]: I don't think you need this constructor param; you can just declare output_buffer with no args.

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:297]: explicit type

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:301]: use std::make_shared rather than raw `new`. Looks like the call-site just assigns this anyway.

File: source/extensions/common/tap/admin.h
Comment (jmarantz) [inline source/extensions/common/tap/admin.h:82]: move variable declarations to a `private:` section at bottom of class decl per Envoy conventions

File: source/extensions/common/tap/admin.h
Comment (jmarantz) [inline source/extensions/common/tap/admin.h:120]: member vars at bottom of class

File: source/extensions/common/tap/admin.h
Comment (jmarantz) [inline source/extensions/common/tap/admin.h:138]: return shared_ptr.

File: test/per_file_coverage.sh
Comment (jmarantz) [inline test/per_file_coverage.sh:34]: can you add tests in this PR rather than slackening our coverage expectations?

**Review by jmarantz — COMMENTED** [review]
This looks great so far -- I left the first round of comments. Haven't looked at the integration test yet.

Can you add unit-tests in addition to the integration test?

File: source/extensions/common/tap/admin.cc
Comment (DavidPeet8) [reply source/extensions/common/tap/admin.cc:162]: I would leave this is as is for a couple reasons. First, I don't think we will have anywhere to send the HTTP response code:
- The response code is already sent at this point because `submitTrace` is called by envoy worker threads, and an HTTP response code is sent when `AdminHandler::handler` completes. 
- Without the attached request, we don't know where to send any of this information

Second, Depending on when the destruction of `attached_request_` occurs, this case may be hit somewhat frequently. For example take the following sequence of events:
1. Worker thread 1 calls `submitTrace` and makes a post to the main thread
2. Worker thread 2 calls `submitTrace` and makes a post to the main thread
3. Main thread executes the post from Worker 1 and finds the buffer is full. The buffer flushes and the stream is closed server side.
4. The onDestroyCallback for the admin stream runs
5. The attached request is deleted
6. The pending post from worker 2 runs, now `admin_request_` does not exist but nothing really went wrong

File: source/extensions/common/tap/admin.cc
Comment (DavidPeet8) [reply source/extensions/common/tap/admin.cc:181]: Unless you feel strongly about this I would lean toward leaving this as is. I have a few reasons for this:

1. Most of the duplicate code lies in the lambda function body which is rather short, and all instances of this duplicate code are very close to one another in the code.
2. I can't imagine Envoy will be introducing many more sink types for the admin handler in the future, so the number of instances of duplicate code should stay fairly low
3. I think we may well end up with more code overall which is more confusing to navigate by adding an intermediary class as it would not have an overly well defined "purpose" other than to avoid code duplication
4. We have to be somewhat careful about what is captured by the lambda with respect to object lifetime. In particular, if I added an intermediate class in the SinkHandle hierarchy, we would need to preserve the lifetime of the sink handle to use its functionality in the lambda function. This is not done by default as a new SinkHandle is created per tapped request.

File: test/per_file_coverage.sh
Comment (DavidPeet8) [reply test/per_file_coverage.sh:34]: The reason why I reduced the coverage bounds is because the uncovered code lies in exceptional cases & race cases. I probably should have added a blurb in the initial PR about this - my bad on that. 

Are there any examples I could follow of how race cases & exceptional cases are tested in Envoy to improve coverage here? I was following the tests for `tap_config_base.cc` but they don't seem to cover the exceptional cases either. For example in `TapConfigBaseImpl::TapConfigBaseImpl` the cases in the `switch` statement causing `PANIC`s are not covered.

Here is the [link](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/102861/artifacts?artifactName=bazel.coverage&api-version=7.0&%24format=zip) to the coverage output for my changes - let me know if there are particular areas you think should be covered better!

**Comment by DavidPeet8** [comment]
I added a unit test, but I'm not sure how much else is unit testable in this change as the types I'm interested in testing are all private inside the `AdminHandler`. Let me know if I'm missing something here

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [reply source/extensions/common/tap/admin.cc:182]: Agreed, lgtm.

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:257]: maybe a slightly more informative variable than 'e1' -- all the reader knows is this is a TraceBuffer, which is evident from context 5 lines above, but I don't know what concept e1 means.

File: source/extensions/common/tap/admin.h
Comment (jmarantz) [inline source/extensions/common/tap/admin.h:77]: consider adding calls to ASSERT_IS_MAIN_OR_TEST_THREAD() from source/common/common/thread.h if you want some asserts about that which disappear in product builds.

File: test/per_file_coverage.sh
Comment (jmarantz) [reply test/per_file_coverage.sh]: Let me take a look at this. the PANICs in switch statements are annoying for sure, and we used to have a better way to exclude lines from coverage.

For reproducing race conditions PTAL @ https://github.com/envoyproxy/envoy/blob/main/source/common/common/thread_synchronizer.h

Comment (jmarantz) [reply test/per_file_coverage.sh]: For covering trivial overrides I think it's fine just to call them, a la https://github.com/envoyproxy/envoy/blob/be8b2594cb5d157e36f39b996a7715812d2e68ae/test/server/admin/admin_test.cc#L253

**Review by jmarantz — COMMENTED** [review]
We probably should get @mattklein123 involved in this as he wrote this filter. But flushing my remaining comments now.

File: source/extensions/common/tap/admin.cc
Comment (DavidPeet8) [reply source/extensions/common/tap/admin.cc:163]: For sure, I'll add some comments with this info

**Comment by DavidPeet8** [comment]
In these changes I added some unit tests for the race cases I mentioned earlier in the review. This increases coverage beyond the original bounds.

I did not end up using the `ThreadSynchronizer` mentioned. After spending some time trying to use the thread synchronizer I came to the following conclusions: 
- Both using the thread synchronizer and directly testing the functions we are interested in require friendship
- Due to the degree of mocking in unit tests, using the thread synchronizer to mirror production code flow would require complex additional code, and would likely both reduce readability and introduce fragile tests. 
- To test these specific cases, all we are really interested in is the order in which posts are made to the main thread dispatcher. I think the cleanest way to test these cases is to explicitly create the dispatcher queue callback ordering, and then let it execute

Let me know your thoughts on the approach I took!

**Comment by DavidPeet8** [comment]
Requesting review as the failing coverage test is due to a directory I have no changes in (`source/common/http`), and the coverage test seems to be failing on many recent commits made to main.

**Comment by mattklein123** [comment]
I think coverage should have been fixed on main. Can you merge main. @jmarantz I think this is ready for another pass? Thank you.

/wait

**Comment by mattklein123** [comment]
@DavidPeet8 friendly request to please not force push. It makes reviews much more difficult. Just merge main and add commits. Thank you!

@jmarantz please ping me when this is ready for my pass. Thanks.

/wait-any

**Comment by DavidPeet8** [comment]
> friendly request to please not force push. It makes reviews much more difficult.

My mistake I didn't realize that!

File: source/extensions/common/tap/admin.cc
Comment (jmarantz) [inline source/extensions/common/tap/admin.cc:140]: nit: `const std::shared_ptr<...>&`

File: source/extensions/common/tap/admin.h
Comment (jmarantz) [inline source/extensions/common/tap/admin.h:92]: nit: const ref.

**Review by jmarantz — DISMISSED** [review]
Thanks for improving the test coverage.

I don't pretend to fully understand the semantics of this, but I'll turn it over to Matt.

File: api/envoy/config/tap/v3/common.proto
Comment (mattklein123) [inline api/envoy/config/tap/v3/common.proto:281]: Can you clarify what happens if there are no traces yet? I assume the endpoint will return with some error code or some message?

File: api/envoy/config/tap/v3/common.proto
Comment (mattklein123) [inline api/envoy/config/tap/v3/common.proto:261]: Please add a release note in current.rst with info about this new feature, and ref link to this field.

File: source/extensions/common/tap/admin.cc
Comment (mattklein123) [inline source/extensions/common/tap/admin.cc:270]: Related to my question above, I guess the answer is if the timer expires and there are no traces, the response will be empty. Should there be some indication that this is a timeout case so the user isn't confused?

File: source/extensions/common/tap/admin.cc
Comment (mattklein123) [inline source/extensions/common/tap/admin.cc:257]: How does this happen if everything is serialized? More comments?

File: source/extensions/common/tap/admin.cc
Comment (mattklein123) [inline source/extensions/common/tap/admin.cc:193]: Some question. How can this happen if serialized?

File: source/extensions/common/tap/admin.cc
Comment (mattklein123) [inline source/extensions/common/tap/admin.cc:137]: Instead of using a switch statement here maybe just use an if statement and assert that type is either admin or streaming admin? Shouldn't the caller have already verified this?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments.

/wait

File: source/extensions/common/tap/admin.cc
Comment (DavidPeet8) [reply source/extensions/common/tap/admin.cc:270]: Yes exactly! How do you feel about adding additional explanation in the proto file as comments? 

I think putting an indicator in the response body would introduce unneeded complexity here as it would make the tests much more difficult. The issue with putting an indicator in the response is that the tests need to be able to deserialize the output as a series of traces in `PROTO_BINARY_LENGTH_DELIMITED` format. If we wanted to add something in the response body we will need to edit the API again by either adding a message or modifying the existing message used.

File: source/extensions/common/tap/admin.cc
Comment (DavidPeet8) [reply source/extensions/common/tap/admin.cc:257]: This case can be hit in the following scenario:
1. The final trace needed to fill the buffer arrives and `submitTrace` posts to the dispatcher.
2. The timeout expires and the on timeout callback is posted to the dispatcher.
3. At this point we have the dispatcher queue with a `submitTrace` post followed by an on timeout post.
4. The callback posted in step 1 runs which flushes the buffer and, posts the on destroy callback for the admin stream. At this point the buffer is flushed but the attached request has not been destroyed.
5. The timeout callback runs, triggering this case.

See `admin_test.cc:BufferedTapDoubleFlush` for the test which covers this case.

Comment (mattklein123) [reply source/extensions/common/tap/admin.cc:257]: Yup OK, can you add more comments? Same in the other case?

**Comment by mattklein123** [comment]
Thanks LGTM but I think this is going to fail format looking at the diff.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by DavidPeet8** [comment]
@mattklein123 the linux_x64 release test seems to be broken on main - is there a way I can get around this or do I need to wait for it to be fixed?

**Comment by mattklein123** [comment]
Sorry it should be fixed by https://github.com/envoyproxy/envoy/pull/20719 so let's wait until that merges and you can merge main.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by DavidPeet8** [comment]
@mattklein123 Are you able to restart the tsan test? As far as I can tell, the failing target `//test/extensions/filters/http/buffer:buffer_filter_integration_test` seems to be unrelated to my changes and failed due to a timeout.

**Comment by jmarantz** [comment]
/retest

@DavidPeet8 FYI anyone can use the "/retest" directive to re-run failed tests.

**Comment by DavidPeet8** [comment]
I didn't realize that, thanks!

**Pull request merged**
